src/lib/transition_handler/catchup_scheduler.ml

```diff
@@ -113,31 +112,39 @@ module Make (Inputs : Inputs.S) = struct
     in
     let make_timeout duration =
       Time.Timeout.create t.time_controller duration ~f:(fun _ ->
-          don't_wait_for (Writer.write t.catchup_job_writer transition) )
+          don't_wait_for (Writer.write t.catchup_job_writer parent_hash) )
     in
-    Hashtbl.update t.collected_transitions parent_hash ~f:(function
-      | None ->
-          let remaining_time = cancel_child_timeout t hash in
-          Hashtbl.add_exn t.collected_transitions ~key:hash ~data:[] ;
-          Hashtbl.add_exn t.parent_root_timeouts ~key:parent_hash
-            ~data:
-              (make_timeout
-                 (Option.value remaining_time ~default:timeout_duration)) ;
-          [transition]
-      | Some sibling_transitions ->
-          if
-            List.exists sibling_transitions ~f:(fun collected_transition ->
-                State_hash.equal hash @@ With_hash.hash collected_transition )
-          then (
-            Logger.info t.logger
-              !"Received request to watch transition for catchup that already \
-                was being watched: %{sexp: State_hash.t}"
-              hash ;
-            sibling_transitions )
-          else
-            let _ : Time.Span.t option = cancel_child_timeout t hash in
-            Hashtbl.add_exn t.collected_transitions ~key:hash ~data:[] ;
-            transition :: sibling_transitions )
+    match Hashtbl.find t.collected_transitions parent_hash with
+    | None ->
+        let remaining_time = cancel_timeout t hash in
+        Hashtbl.add_exn t.collected_transitions ~key:parent_hash
+          ~data:[transition] ;
+        Hashtbl.update t.collected_transitions hash ~f:(function
+          | None -> []
+          | Some ts -> ts ) ;
+        Hashtbl.add_exn t.parent_root_timeouts ~key:parent_hash
+          ~data:
+            (make_timeout
+               (Option.value remaining_time ~default:timeout_duration))
+    | Some sibling_transitions ->
+        if
+          List.exists sibling_transitions ~f:(fun sibling_transition ->
+              External_transition.Verified.equal
+                (With_hash.data transition)
+                (With_hash.data sibling_transition) )
+        then
+          Logger.info t.logger
+            !"Received request to watch transition for catchup that already \
+              was being watched: %{sexp: State_hash.t}"
+            hash
+        else
+          let _ : Time.Span.t option = cancel_timeout t hash in
+          Hashtbl.update t.collected_transitions parent_hash
```

Comment:
With Hashtbl.update, we are trying to find sibling_transition. However, we already found it earlier before. Use. Hashtbl.set instead of update and set the sibling_transitions you already queried.